pkg/ruler/ruler.go

```diff
@@ -535,6 +535,19 @@ func (r *Ruler) listRuleGroupsToSyncForAllUsers(ctx context.Context, reason rule
 }
 
 func (r *Ruler) listRuleGroupsToSyncForUsers(ctx context.Context, userIDs []string, reason rulesSyncReason) (map[string]rulespb.RuleGroupList, error) {
+	// Filter out users which have been explicitly disabled.
+	wIdx := 0
+	for rIdx := 0; rIdx < len(userIDs); rIdx++ {
+		if userID := userIDs[rIdx]; !r.allowedTenants.IsAllowed(userID) {
+			level.Debug(r.logger).Log("msg", "ignoring rule groups for user, not allowed", "user", userID)
+			continue
+		}
+
+		userIDs[wIdx] = userIDs[rIdx]
```

Comment:
This modifies the `userIDs` slice, which may surprise callers of this method.

Instead of doing that, can we add filtering logic to the next `for` loop?

Reply:
I mean... we build `userRings` there, and from that point on only work with that map to see which users to work with.

Reply:
Done: https://github.com/grafana/mimir/pull/5004/commits/6ad069456fb770084cdc5910d604e6df51b93b5d